RFCs/0043-Fn::FindInMap enhancements (default value and intrinsic functions).md

```diff
@@ -0,0 +1,571 @@
+        }
+    },
+    "Resources": {
+        "myEC2Instance": {
+            "Type": "AWS::EC2::Instance",
+            "Properties": {
+                "ImageId": {
+                    "Fn::FindInMap": [
+                        "RegionMap",
+                        { "Fn::If": [ "conditionName", "us-east-1", "us-east-2" ] },
+                        { "Ref": "SecondKey" }
+                    ]
+                },
+                "InstanceType": "t2.small"
+            }
+        }
+    }
+}
+```
+
+## Default Value
+
+Using Fn::FindInMap with the DefaultValue parameter
+
+### YAML
+```yaml
+Mappings:
+  RegionMap:
+    us-east-1:
+      HVM64: ami-0ff8a91507f77f867
+      HVMG2: ami-0a584ac55a7631c0c
+    us-west-1:
+      HVM64: ami-0bdb828fd58c52235
+      HVMG2: ami-066ee5fd4a9ef77f1
+    eu-west-1:
+      HVM64: ami-047bb4163c506cd98
+      HVMG2: ami-0a7c483d527806435
+    ap-southeast-1:
+      HVM64: ami-08569b978cc4dfa10
+      HVMG2: ami-0be9df32ae9f92309
+    ap-northeast-1:
+      HVM64: ami-06cd52961ce9f0d85
+      HVMG2: ami-053cdd503598e4a9d
+Resources:
+  myEC2Instance:
+    Type: AWS::EC2::Instance
+    Properties:
+      ImageId: !FindInMap
+        - RegionMap
+        - us-east-1
+        - not-found
+        - ami-0ff8a91507f77f867
+      InstanceType: t2.small
+```
+
+### JSON
+```json
+{
+  "Mappings": {
+    "RegionMap": {
+      "us-east-1": {
+        "HVM64": "ami-0ff8a91507f77f867",
+        "HVMG2": "ami-0a584ac55a7631c0c"
+      },
+      "us-west-1": {
+        "HVM64": "ami-0bdb828fd58c52235",
+        "HVMG2": "ami-066ee5fd4a9ef77f1"
+      },
+      "eu-west-1": {
+        "HVM64": "ami-047bb4163c506cd98",
+        "HVMG2": "ami-0a7c483d527806435"
+      },
+      "ap-southeast-1": {
+        "HVM64": "ami-08569b978cc4dfa10",
+        "HVMG2": "ami-0be9df32ae9f92309"
+      },
+      "ap-northeast-1": {
+        "HVM64": "ami-06cd52961ce9f0d85",
+        "HVMG2": "ami-053cdd503598e4a9d"
+      }
+    }
+  },
+  "Resources": {
+    "myEC2Instance": {
+      "Type": "AWS::EC2::Instance",
+      "Properties": {
+        "ImageId": {
+          "Fn::FindInMap": [
+            "RegionMap",
+            "us-east-1",
+            "not-found",
+            "ami-0ff8a91507f77f867"
+          ]
+        },
+        "InstanceType": "t2.small"
+      }
+    }
+  }
+}
+```
+
+# Limitations
+
+Users will not be able to nest `Fn::Split` within `Fn::FindInMap` to specify all parameters of `Fn::FindInMap`. Three or four parameters are required in `Fn::FindInMap`.
+The following usage of `Fn::Split` within `Fn::FindInMap` is invalid. 
+
+### YAML
+```yaml
+!FindInMap [ !Split [ ".", "RegionMap.us-east-1.HVM64" ] ]
+```
+
+### JSON
+```json
+"Fn::FindInMap": [ { "Fn::Split": [ ".", "RegionMap.us-east-1.HVM64" ] } ]
+```
+
+If users use a reference to a resource property as a key in `Fn::FindInMap`, the `DefaultValue` will be returned, provided that the `DefaultValue` is supplied and does not reference a resource property.
```

Comment:
This seems confusing, wouldn't it be better to not allow `!Ref Resource`, `!GetAtt` or `!Sub "${Resource}-${Resource.Att}-...."` in `!FindInMap` ?

Reply:
Good point, I've updated RFC to make that clear.